File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/Migrator.java
package de.marmaro.krt.ffupdater;

import android.content.SharedPreferences;

import java.util.Objects;

import de.marmaro.krt.ffupdater.metadata.fetcher.Cache;

public class Migrator {
    public static final String FFUPDATER_VERSION_NAME = "migrator_ffupdater_version_name";
    private final SharedPreferences preferences;
    private final String currentVersion;

    public Migrator(SharedPreferences sharedPreferences) {
        this(sharedPreferences, BuildConfig.VERSION_NAME);
    }

    Migrator(SharedPreferences preferences, String currentVersion) {
        Objects.requireNonNull(preferences);
        Objects.requireNonNull(currentVersion);
        this.preferences = preferences;
        this.currentVersion = currentVersion;
    }

    public void migrate() {
        String lastVersion = preferences.getString(FFUPDATER_VERSION_NAME, "");

        if (!currentVersion.equals(lastVersion)) {
            deleteAvailableMetadataCache();
        }

        preferences.edit()
                .putString(FFUPDATER_VERSION_NAME, currentVersion)
                .apply();
    }

    private void deleteAvailableMetadataCache() {
        final SharedPreferences.Editor editor = preferences.edit();
        preferences.getAll().keySet().stream()
                .filter(key -> key.startsWith(Cache.BASE))
                .forEach(editor::remove);
        editor.apply();
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/Migrator.kt
package de.marmaro.krt.ffupdater

import android.annotation.SuppressLint
import android.content.Context
import androidx.annotation.Keep
import androidx.preference.PreferenceManager
import java.io.File

@Keep
object Migrator {

    private var wasAppUpdated: Boolean? = null

    @SuppressLint("ApplySharedPref")
    fun migrate(context: Context) {
        val preferences = PreferenceManager.getDefaultSharedPreferences(context.applicationContext)
        val lastVersionCode = preferences.getInt(FFUPDATER_VERSION_CODE, 0)

        wasAppUpdated = (lastVersionCode != BuildConfig.VERSION_CODE)

        if (lastVersionCode < 148) { // 78.0.6
            PreferenceManager.getDefaultSharedPreferences(context.applicationContext)
                .edit()
                .remove("lastBackgroundCheckTimestamp")
                .commit()

            File(context.externalCacheDir, "crashlog.txt").delete()
        }

        preferences.edit()
            .putInt(FFUPDATER_VERSION_CODE, BuildConfig.VERSION_CODE)
            .apply()
    }

    fun wasAppUpdated(): Boolean {
        return requireNotNull(wasAppUpdated) { "call migrate() first" }
    }

    private const val FFUPDATER_VERSION_CODE = "migrator_ffupdater_version_code"
}